torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+            invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+    """
+    def __init__(self,
+                 init_scale=2.**24,
+                 growth_factor=1.001,
+                 backoff_factor=0.5,
+                 enabled=True):
+        self._enabled = enabled
+        if enabled:
+            self._init_scale = init_scale
+            # self._scale will be lazily initialized during the first call to scaler.scale(loss or outputs)
+            self._scale = None
+            self._growth_factor = growth_factor
+            self._backoff_factor = backoff_factor
+            self._per_optimizer_states = defaultdict(dict)
+
+    @staticmethod
+    def _scale_not_initialized_error(funcname):
+        return "Attempted to call {} but the scale tensor is None. This may indicate your ".format(funcname) + \
+               "script did not use scaler.scale(loss or outputs) earlier in the iteration."
+
+    def scale(self, outputs):
+        """
+        Multiplies ('scales') a tensor or list of tensors by the scale factor.
+
+        These tensors are typically the outputs of a network. When ``backward()``
+        is called on scaled outputs the resulting gradients will also be scaled.
+
+        Arguments:
+            outputs (Tensor or iterable of Tensors):  Outputs to scale.
+
+        Returns:
+            Scaled outputs.  If this instance of :class:`AmpScaler` is not enabled, outputs are returned unmodified.
+        """
+        if not self._enabled:
+            return outputs
+
+        # Short-circuit for the common case.
+        if isinstance(outputs, torch.Tensor):
+            assert outputs.is_cuda
+            if self._scale is None:
+                self._scale = torch.full((1,), self._init_scale, dtype=torch.float32, device=outputs.device)
+            return outputs * self._scale.to(device=outputs.device, non_blocking=True)
+
+        # Invoke the more complex machinery only if we're treating multiple outputs.
+        stash = [None]  # trick to hold a reference that can be overwritten at any level of the recursion below.
+
+        def apply_scale(val):
+            if isinstance(val, torch.Tensor):
+                assert val.is_cuda
+                if self._scale is None:
+                    self._scale = torch.full((1,), self._init_scale, dtype=torch.float32, device=val.device)
+                if stash[0] is None:
+                    stash[0] = _MultiDeviceReplicator(self._scale)
+                return val * stash[0].get(val.device)
+            elif isinstance(val, container_abcs.Iterable):
+                return type(val)(apply_scale(v) for v in val)
+            else:
+                raise ValueError("outputs must be a Tensor or an iterable of Tensors")
+
+        return apply_scale(outputs)
+
+    def _unscale_grads(self, optimizer, inv_scale, found_inf, allow_fp16=False):
```

Comment:
allow_fp16 is undocumented

Greg: it should not have a default value

Reply:
This function is an implementation detail, it's underscored.  Got rid of default value.